### PR TITLE
[API] orgEditTeam make Fields optional

### PR DESCRIPTION
**Pull request by 6543**
as title

### CODE REVIEW

**Review by zeripath — APPROVED** [review]
We also need team name based team endpoints for the API. (However we'll have to keep giving out the id to not break existing users)

**Review by lunny — APPROVED** [review]

**Review by lunny — CHANGES_REQUESTED** [review]
Oh. This will not work. See `models.UpdateTeam`.

**Comment by 6543** [comment]
@lunny tested and worked!

> https://github.com/go-gitea/gitea/pull/9556/files#diff-3456f19402661bf53eee0c00f4b034c6R206 make sure name is not empty ...

and if you mean https://github.com/go-gitea/gitea/blob/master/models/org_team.go#L581-L586
this only make sure two teams with same name exist ...

**Comment by lunny** [comment]
Have you tested a blank name? Or could you add a test for that situation?

**Comment by 6543** [comment]
@lunny added Test and make all Fields optional

File: integrations/api_team_test.go [hunk 1]
```diff
@@ -86,12 +86,22 @@ func TestAPITeam(t *testing.T) {
 	checkTeamBean(t, apiTeam.ID, teamToEdit.Name, teamToEdit.Description, teamToEdit.IncludesAllRepositories,
 		teamToEdit.Permission, teamToEdit.Units)
 
+	// Edit team without Name
```
Comment (6543) [inline integrations/api_team_test.go]:
```suggestion
	// Edit team Description only
```

**Comment by 6543** [comment]
@lunny I looked throu the whole function and found out that the API need all fields to be filed otherwhies it changes - eaven if you don't intend to :(

now all fields are optional (beside id ...)

File: routers/api/v1/org/team.go [hunk 2]
```diff
@@ -192,17 +192,28 @@ func EditTeam(ctx *context.APIContext, form api.EditTeamOption) {
...
+
+	if form.CanCreateOrgRepo != nil {
+		team.CanCreateOrgRepo = *form.CanCreateOrgRepo
+	}
+
+	if form.Name != "" {
+		team.Name = form.Name
+	}
+
+	if form.Description != "" {
```
Comment (lafriks) [inline routers/api/v1/org/team.go]:
```suggestion
	if len(form.Description) != 0 {
```

File: routers/api/v1/org/team.go [hunk 3]
```diff
@@ -192,17 +192,28 @@ func EditTeam(ctx *context.APIContext, form api.EditTeamOption) {
...
-	team.CanCreateOrgRepo = form.CanCreateOrgRepo
+	if err := team.GetUnits(); err != nil {
+		ctx.InternalServerError(err)
+	}
+
+	if form.CanCreateOrgRepo != nil {
+		team.CanCreateOrgRepo = *form.CanCreateOrgRepo
+	}
+
+	if form.Name != "" {
```
Comment (lafriks) [inline routers/api/v1/org/team.go]:
```suggestion
	if len(form.Name) != 0 {
```

File: routers/api/v1/org/team.go [hunk 4]
```diff
@@ -192,17 +192,28 @@ func EditTeam(ctx *context.APIContext, form api.EditTeamOption) {
...
+	}
+
+	if form.Description != "" {
+		team.Description = form.Description
+	}
 
 	isAuthChanged := false
 	isIncludeAllChanged := false
-	if !team.IsOwnerTeam() {
+	if !team.IsOwnerTeam() && form.Permission != "" {
```
Comment (lafriks) [inline routers/api/v1/org/team.go]:
```suggestion
	if !team.IsOwnerTeam() && len(form.Permission) != 0 {
```

**Review by lafriks — CHANGES_REQUESTED** [review]

**Comment by 6543** [comment]
@lafriks added your suggestions

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
@lunny look's like an xorm issue, can you confirm?

**[E] EditTeam: mssql: Cannot insert explicit value for identity column in table 'team_unit' when IDENTITY_INSERT is set to OFF.**

log:
```log
testlogger.go:71: 2020/01/02 14:14:36 ...ers/routes/routes.go:100:func1() [I] Started PATCH /api/v1/teams/16?token=4b9184a7c8f09c17a73509376e4d8fe22b596e4e for
--
190 | testlogger.go:71: 2020/01/02 14:14:37 ...dules/context/api.go:78:Error() [E] EditTeam: mssql: Cannot insert explicit value for identity column in table 'team_unit' when IDENTITY_INSERT is set to OFF.
191 | api_team_test.go:92:
192 | Error Trace:	integration_test.go:365
193 | integration_test.go:223
194 | api_team_test.go:92
195 | Error:      	Not equal:
196 | expected: 200
197 | actual  : 500
198 | Test:       	TestAPITeam
199 | Messages:   	Request: PATCH /api/v1/teams/16?
```

File: routers/api/v1/org/team.go [hunk 5]
```diff
@@ -192,17 +192,28 @@ func EditTeam(ctx *context.APIContext, form api.EditTeamOption) {
 	//     "$ref": "#/responses/Team"
 
 	team := ctx.Org.Team
-	team.Description = form.Description
-	unitTypes := models.FindUnitTypes(form.Units...)
-	team.CanCreateOrgRepo = form.CanCreateOrgRepo
+	if err := team.GetUnits(); err != nil {
+		ctx.InternalServerError(err)
```
Comment (lunny) [inline routers/api/v1/org/team.go:196]: return

File: routers/api/v1/org/team.go [hunk 6]
```diff
@@ -192,17 +192,28 @@ func EditTeam(ctx *context.APIContext, form api.EditTeamOption) {
...
+
+	if form.CanCreateOrgRepo != nil {
+		team.CanCreateOrgRepo = *form.CanCreateOrgRepo
+	}
+
+	if len(form.Name) != 0 {
+		team.Name = form.Name
+	}
+
+	if len(form.Description) != 0 {
```
Comment (lunny) [inline routers/api/v1/org/team.go]: So how to set the description as blank?

File: routers/api/v1/org/team.go [hunk 7]
```diff
@@ -215,14 +226,17 @@ func EditTeam(ctx *context.APIContext, form api.EditTeamOption) {
...
+					OrgID: ctx.Org.Team.OrgID,
+					Type:  tp,
+				})
+			}
+			team.Units = units
 		}
-		team.Units = units
 	}
 
 	if err := models.UpdateTeam(team, isAuthChanged, isIncludeAllChanged); err != nil {
```
Comment (lunny) [inline routers/api/v1/org/team.go:243]: But this function will update all fields of the struct?

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
@6543 You should not insert records with id on mssql since it's autoincrement. But mysql will support that.

File: routers/api/v1/org/team.go (hunk 6 above)
Comment (6543) [reply routers/api/v1/org/team.go]: @lunny good point I'll change it as i did with `CanCreateOrgRepo`

File: routers/api/v1/org/team.go (hunk 7 above)
Comment (6543) [reply routers/api/v1/org/team.go:243]: @lunny yes it does --- thats how the function wors NOW

**Comment by 6543** [comment]
@lunny done

https://github.com/go-gitea/gitea/pull/9556/files#diff-f13bb9abff87d33ca9cfc067797eb48bL609-R609

**Review by lunny — APPROVED** [review]

**Pull request merged**